.bazelrc

```diff
@@ -177,6 +178,7 @@ build:clang-msan --copt -fsanitize=memory
 build:clang-msan --linkopt -fsanitize=memory
 build:clang-msan --linkopt -fuse-ld=lld
 build:clang-msan --copt -fsanitize-memory-track-origins=2
+build:clang-msan --copt -fno-sanitize-memory-use-after-dtor
```

Comment:
cc @RyanTheOptimist 

these are the tests that are failing:

```console
//test/common/http:conn_manager_impl_test                                FAILED in 163.2s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/http/conn_manager_impl_test/test.log
//test/common/http/http3:conn_pool_test                                  FAILED in 2.6s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/http/http3/conn_pool_test/test.log
//test/common/listener_manager:listener_manager_impl_test                FAILED in 215.5s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/listener_manager/listener_manager_impl_test/test.log
//test/common/quic:client_connection_factory_impl_test                   FAILED in 2.7s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/quic/client_connection_factory_impl_test/test.log
//test/common/quic:quic_transport_socket_factory_test                    FAILED in 3.3s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/quic/quic_transport_socket_factory_test/test.log
//test/common/router:config_impl_test                                    FAILED in 171.0s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/router/config_impl_test/test.log
//test/common/router:router_test                                         FAILED in 165.9s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/router/router_test/test.log
//test/common/stats:thread_local_store_test                              FAILED in 16.0s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/stats/thread_local_store_test/test.log
//test/common/tls:ssl_socket_test                                        FAILED in 256.7s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/tls/ssl_socket_test/test.log
//test/common/upstream:cluster_manager_impl_test                         FAILED in 189.5s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/common/upstream/cluster_manager_impl_test/test.log
//test/config_test:example_configs_test                                  FAILED in 69.4s
  /build/bazel_root/base/execroot/envoy/bazel-out/k8-dbg/testlogs/test/config_test/example_configs_test/test.log
//test/integration:alpn_integration_test                                 FAILED in 2 out of 2 in 6.0s
  Stats over 2 runs: max = 6.0s, min = 6.0s, avg = 6.0s, dev = 0.0s
```

this is one of the issues:

```console
[ RUN      ] StatsThreadLocalStoreTestNoFixture.MemoryWithTlsRealSymbolTable
==16==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55f6dab3d04c in Envoy::ThreadLocal::MockInstance::SlotImpl::~SlotImpl() /proc/self/cwd/test/mocks/thread_local/mocks.h:59:30
    #1 0x55f6dab3d12d in Envoy::ThreadLocal::MockInstance::SlotImpl::~SlotImpl() /proc/self/cwd/test/mocks/thread_local/mocks.h:55:26
    #2 0x55f6da630139 in Envoy::Stats::ThreadLocalStoreImpl::~ThreadLocalStoreImpl() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:65:5
    #3 0x55f6da529934 in Envoy::Stats::StatsThreadLocalStoreTestNoFixture::~StatsThreadLocalStoreTestNoFixture() /proc/self/cwd/test/common/stats/thread_local_store_test.cc:1505:3
    #4 0x55f6da5299dd in Envoy::Stats::StatsThreadLocalStoreTestNoFixture_MemoryWithTlsRealSymbolTable_Test::~StatsThreadLocalStoreTestNoFixture_MemoryWithTlsRealSymbolTable_Test() /proc/self/cwd/test/common/stats/thread_local_store_test.cc:1533:1
    #5 0x55f6df35025c in testing::Test::DeleteSelf_() /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:500:24
    #6 0x55f6df34f560 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #7 0x55f6df30934e in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2838:5
    #8 0x55f6df30b0a1 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #9 0x55f6df33eae3 in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #10 0x55f6df351b20 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #11 0x55f6df33de3d in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #12 0x55f6dd092ef5 in Envoy::TestRunner::runTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #13 0x55f6dd08ea8e in main /proc/self/cwd/test/main.cc:34:10
    #14 0x7fe0aa84e082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: eebe5d5f4b608b8a53ec446b63981bba373ca0ca)
    #15 0x55f6da38ee9d in _start (/b/f/w/bazel-out/k8-dbg/bin/test/common/stats/thread_local_store_test.runfiles/envoy/test/common/stats/thread_local_store_test+0x128ae9d)

  Member fields were destroyed
    #0 0x55f6da3c266d in __sanitizer_dtor_callback_fields /tmp/llvm_release/final/llvm-project/compiler-rt/lib/msan/msan_interceptors.cpp:961:5
    #1 0x55f6dab38528 in Envoy::ThreadLocal::MockInstance::~MockInstance() /proc/self/cwd/test/mocks/thread_local/mocks.h:107:8
    #2 0x55f6da5298b3 in Envoy::Stats::StatsThreadLocalStoreTestNoFixture::~StatsThreadLocalStoreTestNoFixture() /proc/self/cwd/external/com_google_googletest/googlemock/include/gmock/gmock-nice-strict.h:103:3
    #3 0x55f6da5299dd in Envoy::Stats::StatsThreadLocalStoreTestNoFixture_MemoryWithTlsRealSymbolTable_Test::~StatsThreadLocalStoreTestNoFixture_MemoryWithTlsRealSymbolTable_Test() /proc/self/cwd/test/common/stats/thread_local_store_test.cc:1533:1
    #4 0x55f6df35025c in testing::Test::DeleteSelf_() /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:500:24
    #5 0x55f6df34f560 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #6 0x55f6df30934e in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2838:5
    #7 0x55f6df30b0a1 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #8 0x55f6df33eae3 in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #9 0x55f6df351b20 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #10 0x55f6df33de3d in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #11 0x55f6dd092ef5 in Envoy::TestRunner::runTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #12 0x55f6dd08ea8e in main /proc/self/cwd/test/main.cc:34:10
    #13 0x7fe0aa84e082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: eebe5d5f4b608b8a53ec446b63981bba373ca0ca)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /proc/self/cwd/test/mocks/thread_local/mocks.h:59:30 in Envoy::ThreadLocal::MockInstance::SlotImpl::~SlotImpl()
```

to my untrained eye it looks similar (mockingly) to another issue:

```console
==16==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55d3b1b09ff8 in Envoy::Upstream::ClusterManagerImpl::isShutdown() /proc/self/cwd/./source/common/upstream/cluster_manager_impl.h:314:32
    #1 0x55d3b44031fd in Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster::~Cluster() /proc/self/cwd/source/extensions/clusters/dynamic_forward_proxy/cluster.cc:85:11
    #2 0x55d3b440499d in Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster::~Cluster() /proc/self/cwd/source/extensions/clusters/dynamic_forward_proxy/cluster.cc:80:21
    #3 0x55d3b44373d4 in std::__1::__shared_ptr_pointer<Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster*, std::__1::shared_ptr<Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster>::__shared_ptr_default_delete<Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster, Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster>, std::__1::allocator<Envoy::Extensions::Clusters::DynamicForwardProxy::Cluster>>::__on_zero_shared() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:65:5
    #4 0x55d3b559d549 in Envoy::Upstream::ClusterManagerImpl::ClusterData::~ClusterData() /opt/llvm/bin/../include/c++/v1/__memory/shared_ptr.h:157:9
    #5 0x55d3b559d69d in Envoy::Upstream::ClusterManagerImpl::ClusterData::~ClusterData() /proc/self/cwd/./source/common/upstream/cluster_manager_impl.h:737:10
    #6 0x55d3b1b0be81 in std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::unique_ptr<Envoy::Upstream::ClusterManagerImpl::ClusterData, std::__1::default_delete<Envoy::Upstream::ClusterManagerImpl::ClusterData>>>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::unique_ptr<Envoy::Upstream::ClusterManagerImpl::ClusterData, std::__1::default_delete<Envoy::Upstream::ClusterManagerImpl::ClusterData>>>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::unique_ptr<Envoy::Upstream::ClusterManagerImpl::ClusterData, std::__1::default_delete<Envoy::Upstream::ClusterManagerImpl::ClusterData>>>>>::destroy(std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::unique_ptr<Envoy::Upstream::ClusterManagerImpl::ClusterData, std::__1::default_delete<Envoy::Upstream::ClusterManagerImpl::ClusterData>>>, void*>*) /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:65:5
    #7 0x55d3b1b08056 in Envoy::Upstream::ClusterManagerImpl::~ClusterManagerImpl() /opt/llvm/bin/../include/c++/v1/__tree:1801:3
    #8 0x55d3b1b0821d in Envoy::Upstream::ValidationClusterManager::~ValidationClusterManager() /proc/self/cwd/./source/server/config_validation/cluster_manager.h:45:7
    #9 0x55d3b1a17afd in Envoy::Server::Configuration::MainImpl::~MainImpl() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:65:5
    #10 0x55d3b19fc1b2 in Envoy::ConfigTest::ConfigTest::ConfigTest(Envoy::OptionsImpl&) /proc/self/cwd/test/config_test/config_test.cc:168:3
    #11 0x55d3b19e7865 in Envoy::ConfigTest::run(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /proc/self/cwd/test/config_test/config_test.cc:244:18
    #12 0x55d3b1986471 in Envoy::ExampleConfigsTest_All_Test::TestBody() /proc/self/cwd/test/config_test/example_configs_test.cc:30:3
    #13 0x55d3c22ee5e0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #14 0x55d3c22a5240 in testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2655:5
    #15 0x55d3c22a84d1 in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2832:11
    #16 0x55d3c22aa2f1 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #17 0x55d3c22ddd33 in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #18 0x55d3c22f0ba0 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #19 0x55d3c22dd08d in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #20 0x55d3c013cf15 in Envoy::TestRunner::runTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #21 0x55d3c0138aae in main /proc/self/cwd/test/main.cc:34:10
    #22 0x7f8e917a8082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: eebe5d5f4b608b8a53ec446b63981bba373ca0ca)
    #23 0x55d3b18fe02d in _start (/b/f/w/bazel-out/k8-dbg/bin/test/config_test/example_configs_test.runfiles/envoy/test/config_test/example_configs_test+0x361502d)

  Member fields were destroyed
    #0 0x55d3b19317fd in __sanitizer_dtor_callback_fields /tmp/llvm_release/final/llvm-project/compiler-rt/lib/msan/msan_interceptors.cpp:961:5
    #1 0x55d3b1b077ab in Envoy::Upstream::ClusterManagerImpl::~ClusterManagerImpl() /proc/self/cwd/./source/common/upstream/cluster_manager_impl.h:937:8
    #2 0x55d3b1b0821d in Envoy::Upstream::ValidationClusterManager::~ValidationClusterManager() /proc/self/cwd/./source/server/config_validation/cluster_manager.h:45:7
    #3 0x55d3b1a17afd in Envoy::Server::Configuration::MainImpl::~MainImpl() /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:65:5
    #4 0x55d3b19fc1b2 in Envoy::ConfigTest::ConfigTest::ConfigTest(Envoy::OptionsImpl&) /proc/self/cwd/test/config_test/config_test.cc:168:3
    #5 0x55d3b19e7865 in Envoy::ConfigTest::run(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) /proc/self/cwd/test/config_test/config_test.cc:244:18
    #6 0x55d3b1986471 in Envoy::ExampleConfigsTest_All_Test::TestBody() /proc/self/cwd/test/config_test/example_configs_test.cc:30:3
    #7 0x55d3c22ee5e0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #8 0x55d3c22a5240 in testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2655:5
    #9 0x55d3c22a84d1 in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2832:11
    #10 0x55d3c22aa2f1 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #11 0x55d3c22ddd33 in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #12 0x55d3c22f0ba0 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #13 0x55d3c22dd08d in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #14 0x55d3c013cf15 in Envoy::TestRunner::runTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #15 0x55d3c0138aae in main /proc/self/cwd/test/main.cc:34:10
    #16 0x7f8e917a8082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: eebe5d5f4b608b8a53ec446b63981bba373ca0ca)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /proc/self/cwd/./source/common/upstream/cluster_manager_impl.h:314:32 in Envoy::Upstream::ClusterManagerImpl::isShutdown()

```

... at least in the sense that its borking on a variable named `shutdown_` after the object is destructed